File: src/leetcode/python/easy/NthDigit_400.py
class NthDigit_400(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        ln, bs = 1, 1
        while 9 * ln * bs < n:
            n -= 9 * ln * bs
            bs *= 10
            ln += 1
        ans = 0
        num = (n - 1) // ln + bs
        i = (n - 1) % ln
        while i < ln:
            ans = num % 10
            num //= 10
            i += 1
        return ans

File: src/leetcode/java/NthDigit_400.java
package leetcode.java.easy;

public class NthDigit_400 {
  public int findNthDigit(int n) {
    int len = 1, base = 1;
    for (; n > 9L * base * len; base *= 10, len++) {
      n -= 9 * base * len;
    }
    int ans = 0, num = (n - 1) / len + base;
    for (int i = (n - 1) % len; i < len; i++) {
      ans = num % 10;
      num /= 10;
    }
    return ans;
  }
}
